File: templates/objc/metal_manager.m
#import "metal_manager.h"

@implementation MetalManager {
    id<MTLDevice> _device;
    id<MTLCommandQueue> _commandQueue;
}

- (instancetype)initWithDevice:(id<MTLDevice>)device {
    self = [super init];
    if (self) {
        _device = device;
        _commandQueue = [_device newCommandQueue];
    }
    return self;
}

- (void)executeKernelWithName:(NSString *)kernelName
                    withInput:(id<MTLBuffer>)inputBuffer
                   outputBuffer:(id<MTLBuffer>)outputBuffer {
    NSError *error = nil;
    id<MTLLibrary> library = [_device newDefaultLibrary];
    id<MTLFunction> function = [library newFunctionWithName:kernelName];

    if (!function) {
        NSLog(@"Failed to load kernel function: %@", kernelName);
        return;
    }

    id<MTLComputePipelineState> pipelineState = [_device newComputePipelineStateWithFunction:function error:&error];
    if (error) {
        NSLog(@"Error creating pipeline state: %@", error.localizedDescription);
        return;
    }

    id<MTLCommandBuffer> commandBuffer = [_commandQueue commandBuffer];
    id<MTLComputeCommandEncoder> commandEncoder = [commandBuffer computeCommandEncoder];

    [commandEncoder setComputePipelineState:pipelineState];
    [commandEncoder setBuffer:inputBuffer offset:0 atIndex:0];
    [commandEncoder setBuffer:outputBuffer offset:0 atIndex:1];

    MTLSize gridSize = MTLSizeMake(256, 1, 1);
    MTLSize threadGroupSize = MTLSizeMake(16, 1, 1);
    [commandEncoder dispatchThreads:gridSize threadsPerThreadgroup:threadGroupSize];

    [commandEncoder endEncoding];
    [commandBuffer commit];
    [commandBuffer waitUntilCompleted];

    NSLog(@"Kernel execution complete.");
}

@end

File: templates/swift/metal_manager.swift
import Metal
import Foundation

class MetalManager {
    private let device: MTLDevice
    private let commandQueue: MTLCommandQueue

    init(device: MTLDevice) {
        self.device = device
        self.commandQueue = device.makeCommandQueue()!
    }

    func executeKernel(functionName: String, inputBuffer: MTLBuffer, outputBuffer: MTLBuffer) {
        guard let library = device.makeDefaultLibrary(),
              let function = library.makeFunction(name: functionName) else {
            print("Failed to find the function \(functionName)")
            return
        }

        do {
            let pipelineState = try device.makeComputePipelineState(function: function)
            guard let commandBuffer = commandQueue.makeCommandBuffer(),
                  let commandEncoder = commandBuffer.makeComputeCommandEncoder() else {
                print("Failed to create command encoder")
                return
            }

            commandEncoder.setComputePipelineState(pipelineState)
            commandEncoder.setBuffer(inputBuffer, offset: 0, index: 0)
            commandEncoder.setBuffer(outputBuffer, offset: 0, index: 1)

            let gridSize = MTLSize(width: 256, height: 1, depth: 1)
            let threadGroupSize = MTLSize(width: 16, height: 1, depth: 1)
            commandEncoder.dispatchThreads(gridSize, threadsPerThreadgroup: threadGroupSize)

            commandEncoder.endEncoding()
            commandBuffer.commit()
            commandBuffer.waitUntilCompleted()

            print("Kernel execution completed")
        } catch {
            print("Error creating pipeline state: \(error)")
        }
    }
}
